### PR TITLE
Mark GC blocked when acquiring mutex

### PR DESCRIPTION
**Pull request by ugexe**
Fixes: `perl6 -e 'my $foo = 0; for ^50000 { start { say $foo++ } };'`

https://irclog.perlgeek.de/perl6-dev/2017-09-06#i_15127640
https://rt.perl.org/Ticket/Display.html?id=132042

### CODE REVIEW

File: src/io/io.c [hunk 1]
```diff
@@ -157,10 +177,13 @@ MVMObject * MVM_io_read_bytes_async(MVMThreadContext *tc, MVMObject *oshandle, M
...
     if (handle->body.ops->async_readable) {
-        uv_mutex_t *mutex = acquire_mutex(tc, handle);
-        MVMObject *result = (MVMObject *)handle->body.ops->async_readable->read_bytes(tc,
-            handle, queue, schedulee, buf_type, async_type);
-        release_mutex(tc, mutex);
+        MVMObject *result;
+        MVMROOT(tc, handle, {
+            uv_mutex_t *mutex = acquire_mutex(tc, handle);
+            result = (MVMObject *)handle->body.ops->async_readable->read_bytes(tc,
+                handle, queue, schedulee, buf_type, async_type);
```
Comment (niner) [inline src/io/io.c]: queue, schedulee, buf_type and async_type must also be MVMROOTed.
All pointers to garbage collected objects (like MVMObject and subtypes) that you access after a GC triggering function (acquire_mutex here) must be rooted. Doesn't matter if you use them within the MVMROOT block or after it. The MVMROOT block is the part where GC may be triggered and thus the pointer may have to be updated.

File: src/io/io.c [hunk 2]
```diff
@@ -173,10 +196,13 @@ MVMObject * MVM_io_write_bytes_async(MVMThreadContext *tc, MVMObject *oshandle,
...
     if (handle->body.ops->async_writable) {
-        uv_mutex_t *mutex = acquire_mutex(tc, handle);
-        MVMObject *result = (MVMObject *)handle->body.ops->async_writable->write_bytes(tc,
-            handle, queue, schedulee, buffer, async_type);
-        release_mutex(tc, mutex);
+        MVMObject *result;
+        MVMROOT(tc, handle, {
+            uv_mutex_t *mutex = acquire_mutex(tc, handle);
+            result = (MVMObject *)handle->body.ops->async_writable->write_bytes(tc,
+                handle, queue, schedulee, buffer, async_type);
```
Comment (niner) [inline src/io/io.c]: queue, schedulee, buffer, async_type need rooting like above.

File: src/io/io.c [hunk 3]
```diff
@@ -190,10 +216,13 @@ MVMObject * MVM_io_write_bytes_to_async(MVMThreadContext *tc, MVMObject *oshandl
...
     if (handle->body.ops->async_writable_to) {
-        uv_mutex_t *mutex = acquire_mutex(tc, handle);
-        MVMObject *result = (MVMObject *)handle->body.ops->async_writable_to->write_bytes_to(tc,
-            handle, queue, schedulee, buffer, async_type, host, port);
-        release_mutex(tc, mutex);
+        MVMObject *result;
+        MVMROOT(tc, handle, {
+            uv_mutex_t *mutex = acquire_mutex(tc, handle);
+            result = (MVMObject *)handle->body.ops->async_writable_to->write_bytes_to(tc,
+                handle, queue, schedulee, buffer, async_type, host, port);
```
Comment (niner) [inline src/io/io.c]: Again queue, schedulee, buffer, async_type and host need rooting.

File: src/io/io.c [hunk 4]
```diff
@@ -260,9 +301,11 @@ void MVM_io_truncate(MVMThreadContext *tc, MVMObject *oshandle, MVMint64 offset)
 void MVM_io_connect(MVMThreadContext *tc, MVMObject *oshandle, MVMString *host, MVMint64 port) {
     MVMOSHandle *handle = verify_is_handle(tc, oshandle, "connect");
     if (handle->body.ops->sockety) {
-        uv_mutex_t *mutex = acquire_mutex(tc, handle);
-        handle->body.ops->sockety->connect(tc, handle, host, port);
-        release_mutex(tc, mutex);
+        MVMROOT(tc, handle, {
+            uv_mutex_t *mutex = acquire_mutex(tc, handle);
+            handle->body.ops->sockety->connect(tc, handle, host, port);
```
Comment (niner) [inline src/io/io.c]: host needs rooting

File: src/io/io.c [hunk 5]
```diff
@@ -271,9 +314,11 @@ void MVM_io_connect(MVMThreadContext *tc, MVMObject *oshandle, MVMString *host,
 void MVM_io_bind(MVMThreadContext *tc, MVMObject *oshandle, MVMString *host, MVMint64 port, MVMint32 backlog) {
     MVMOSHandle *handle = verify_is_handle(tc, oshandle, "bind");
     if (handle->body.ops->sockety) {
-        uv_mutex_t *mutex = acquire_mutex(tc, handle);
-        handle->body.ops->sockety->bind(tc, handle, host, port, backlog);
-        release_mutex(tc, mutex);
+        MVMROOT(tc, handle, {
+            uv_mutex_t *mutex = acquire_mutex(tc, handle);
+            handle->body.ops->sockety->bind(tc, handle, host, port, backlog);
```
Comment (niner) [inline src/io/io.c]: host needs rooting

File: src/io/io.c (hunk 1 above)
Comment (ugexe) [reply src/io/io.c]: Thanks for being explicit about the what and why. Will attempt to address later today.

**Pull request merged**